packages/extra/src/LocalDateRange.js

```diff
@@ -0,0 +1,742 @@
+ * No empty range can exist at `LocalDate.MIN` or `LocalDate.MAX`.
+ * 
+ * Date ranges are not comparable. To compare the length of two ranges, it is
+ * generally recommended to compare the number of days they contain.
+ *
+ */
+export class LocalDateRange {
+    /**
+     * function overloading for {@link LocalDateRange.of}
+     *
+     * if called with `LocalDate` and `LocalDate`, {@link LocalDateRange._ofLocalDateLocalDate} is executed,
+     * if called with `LocalDate` and `Period`, {@link LocalDateRange._ofLocalDatePeriod} is executed,
+     * otherwise throws IllegalArgumentException.
+     *
+     * @param {LocalDate} startInclusive
+     * @param {LocalDate|Period} endExclusiveOrPeriod
+     * @returns {LocalDateRange}
+     */
+    static of(startInclusive, endExclusiveOrPeriod) {
+        if (startInclusive instanceof LocalDate && endExclusiveOrPeriod instanceof LocalDate) {
+            return LocalDateRange._ofLocalDateLocalDate(startInclusive, endExclusiveOrPeriod);
+        }
+        if (startInclusive instanceof LocalDate && endExclusiveOrPeriod instanceof Period) {
+            return LocalDateRange._ofLocalDatePeriod(startInclusive, endExclusiveOrPeriod);
+        }
+        // FIXME
+        const startInclusiveMessage = `startInclusive must be an instance of LocalDate but is ${startInclusive.constructor.name}`;
+        const endExclusiveOrPeriodMessage = `endExclusiveOrPeriod must be an instance of LocalDate or Period but is ${endExclusiveOrPeriod.constructor.name}`;
+        throw new IllegalArgumentException(`${startInclusiveMessage} and ${endExclusiveOrPeriodMessage}`);
+    }
+
+    //-----------------------------------------------------------------------
+    /**
+     * Obtains a half-open range of dates, including the start and excluding the end.
+     * 
+     * The range includes the start date and excludes the end date, unless the end is `LocalDate.MAX`.
+     * The end date must be equal to or after the start date.
+     * This definition permits an empty range located at a specific date.
+     * 
+     * The constants `LocalDate.MIN` and `LocalDate.MAX` can be used
+     * to indicate an unbounded far-past or far-future.
+     * 
+     * The start inclusive date must not be `LocalDate.MAX` or `LocalDate.MAX.minusDays(1)`.
+     * The end inclusive date must not be `LocalDate.MIN` or `LocalDate.MIN.plusDays(1)`.
+     * No empty range can exist at `LocalDate.MIN` or `LocalDate.MAX`.
+     *
+     * @param {LocalDate} startInclusive  the inclusive start date, not null
+     * @param {LocalDate} endExclusive  the exclusive end date, not null
+     * @return {LocalDateRange} the half-open range, not null
+     * @throws DateTimeException if the end is before the start,
+     *   or the start date is `LocalDate.MAX` or `LocalDate.MAX.minusDays(1)`,
+     *   or the end date is `LocalDate.MIN` or `LocalDate.MIN.plusDays(1)`
+     */
+    static _ofLocalDateLocalDate(startInclusive, endExclusive) {
```

Comment:
What do you mean by "excluded from the docs"?

As for `@Private`, I would be more than happy to change almost all `_` to [`#`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) - probably with the exception of protected-like methods needed by `Temporal`.

In the meanwhile, I'll try to find if it's possible in JS to use template method patterns and support encapsulation at the same time.

Reply:
Its more like a convention. all internal methods, that should not be used from "outside" and are not part of the API, should start with an `_` and have the `@private` annotation. The `@private` annotation excludes it from the generated esdocs. The `_` makes it obvious for developers, that are inspecting the code.
I am not sure if we fulfill this convention everywhere in our code base, but the most places in our `core` should.

Reply:
As for docs, I'm unable to force JSDoc to document more than one overload.

```javascript
/**
 * (...)
 * @param {LocalDate} date (...)
 * @param {ZoneOffset} offset (...)
 * @return {OffsetDate} (...)
 *//**
 * (...)
 * @param {number} year (...)
 * @param {number} month (...)
 * @param {number} dayOfMonth (...)
 * @param {ZoneOffset} offset (...)
 * @return {OffsetDate} (...)
 * @throws DateTimeException (...)
 */
static of() {
    switch (arguments.length) {
        case 2: return OffsetDate._ofLocalDateZoneOffset(...arguments);
        case 4: return OffsetDate._ofIntIntIntZoneOffset(...arguments);
        default: throw new IllegalArgumentException('Illegal number of arguments');
    }
}
```

Building esdoc ends up with:
```console
warning: signature mismatch: OffsetDate.of
```

It's also worth nothing, that with this ugly pattern there is no way to document `IllegalArgumentException` thrown from dispatching method itself.
On the other hand, making `OffsetDate._of*` public just for the sake of documentation doesn't feel right.

@pithu, @InExtremaRes Any ideas?

Reply:
I just realized JSDoc isn't working properly because... it's not used at all.
Are there any plans for migrating to it from already dead ESDoc?

Reply:
No, we don't have any plans to migrate away from esdoc.

Reply:
Running `npm run build-esdoc` works just fine for me, so i might not understand your issue.

I see that there are still some `now0()` methods not declared private. What about change the name to `_now0()` ?

Beside the issue with Locale mentioned by @InExtremaRes, all the rest looks good to me now.

Reply:
This makes documenting of`OffsetDate#of()` virtually impossible. Either `OffsetDate._of*`have to be public (which is ugly) or ESDoc supports overloads (which won't happen).

Reply:
> Running `npm run build-esdoc` works just fine for me, so i might not understand your issue.

Either
Change `OffsetDate.of()` docs as in https://github.com/js-joda/js-joda/pull/622#discussion_r960829833. ESDoc doesn't support multiple `/** ... */`, so it will document just one of overload.

Or
Leave `OffsetDate.of()` as is, but make `ofLocalDateZoneOffset` and `ofIntIntIntZoneOffset` `@private`. This way documentation becomes useless, as there won't be anything to link to.

```
     * if called with 2 arguments {@link OffsetDate.ofLocalDateZoneOffset} is executed,
     * if called with 4 agruments {@link YearQuarter.ofIntIntIntZoneOffset} is executed,
```

BTW YearQuarter.ofIntIntIntZoneOffset should be changed to OffsetDate.ofIntIntIntZoneOffset.

Reply:
For me its fine like it is with `OffsetDate.of()`, the typescript typing showing the correct API and to have this "ugly" public methods like `ofIntIntIntZoneOffset` in the js API is not nice, but the limitation we have when migration from a language with function overloading

But why not fixing the `now0` method ? and some private constructors are still in the esdoc, these ones could be removed from esdoc with the `@private` annotation.

In summary, where possible we should have the docs aligned with the typescript definitions and where not possible we have to live with some work arounds.

Reply:
> For me its fine like it is with `OffsetDate.of()`, the typescript typing showing the correct API and to have this "ugly" public methods like `ofIntIntIntZoneOffset` in the js API is not nice, but the limitation we have when migration from a language with function overloading.

It's just because of ESDoc. We could have no "ugly" public methods at all. Maybe migrating away from ESDoc is worth reconsidering? As a last resort I'll rename `ofIntIntIntZoneOffset` to `_ofIntIntIntZoneOffset`, but it still will have to public, probably forever.

> But i stumbled over some private constructors that are still in the esdoc, these ones could be removed from esdoc with the @private annotation. eg the OffsetDate constructor

Sure, fixed this already but didn't push because I was waiting for `of*()`-related discussion to be resolved.